soroban-env-host/src/e2e_invoke.rs

```diff
@@ -418,9 +420,25 @@ fn build_storage_map_from_xdr_ledger_entries<T: AsRef<[u8]>, I: ExactSizeIterato
                 budget,
             )?;
 
+            if ee.live_until_ledger_seq < ledger_num {
```

Comment:
This is fine for now, but I think we need to figure out how to enforce this at the storage layer instead. Quite a few storage functions depend on the entries being never expired, so it would be better to enforce it there.

Reply:
This check is done before the host is created. What's the benefit to moving it into storage? Checking that we aren't somehow reducing expiration or messing with the ledger_num?

Reply:
> What's the benefit to moving it into storage?

Yeah, storage is probably not the right place, because it doesn't know about the ledger state. But host constructor probably is (see the second message).

> Checking that we aren't somehow reducing expiration or messing with the ledger_num?

The benefit is to enforce the invariant in a more appropriate place. `e2e_invoke` is just once specific code path that creates host. However, host/storage are written in a way that assumes that this invariant always holds. So it would make more sense to enforce that invariant at least during the host creation. Besides the benefit of having more clear expectations and reducing maintenance risks, this also would bring the logic to the preflight (which currently doesn't use `e2e_invoke`, but does support preflight with user-specified input).

Reply:
Hmm yeah that's not a bad idea. So update `with_storage_and_budget` to return a Result or HostError and do the validation in there.

Reply:
I'm not sure how disruptive of a change it'll be downstream to allow errors there though.

Reply:
It will indeed be somewhat disruptive, so we might skip this for now. Also, I've just realized that preflight doesn't do enforcing footprint, so it won't be as beneficial (I confused that with enforcing auth that they can do).